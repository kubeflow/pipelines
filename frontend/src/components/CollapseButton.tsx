/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { CompareState } from '../pages/Compare';
import Button from '@material-ui/core/Button';
import ExpandedIcon from '@material-ui/icons/ArrowDropUp';
import { stylesheet, classes } from 'typestyle';
import { color, fontsize } from '../Css';

const css = stylesheet({
  collapseBtn: {
    color: color.strong,
    fontSize: fontsize.title,
    fontWeight: 'bold',
    padding: 5,
  },
  collapsed: {
    transform: 'rotate(180deg)',
  },
  icon: {
    marginRight: 5,
    transition: 'transform 0.3s',
  },
});

interface CollapseButtonProps {
  collapseSections: { [key: string]: boolean };
  compareSetState: (state: Partial<CompareState>) => void;
  sectionName: string;
}

class CollapseButton extends React.Component<CollapseButtonProps> {
  public render(): JSX.Element {
    const { collapseSections, compareSetState } = this.props;
    const sectionName = this.props.sectionName;
    return (
      <div>
        <Button
          onClick={() => {
            collapseSections[sectionName] = !collapseSections[sectionName];
            compareSetState({ collapseSections });
          }}
          title='Expand/Collapse this section'
          className={css.collapseBtn}
        >
          <ExpandedIcon
            className={classes(css.icon, collapseSections[sectionName] ? css.collapsed : '')}
          />
          {sectionName}
        </Button>
      </div>
    );
  }
}

export default CollapseButton;
