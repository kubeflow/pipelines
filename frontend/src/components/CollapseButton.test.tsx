/*
 * Copyright 2018 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import CollapseButton from './CollapseButton';
import { shallow } from 'enzyme';

describe('CollapseButton', () => {
  const compareComponent = {
    setState: jest.fn(),
    state: {
      collapseSections: {},
    },
  } as any;

  afterEach(() => (compareComponent.state.collapseSections = {}));

  it('initial render', () => {
    const tree = shallow(
      <CollapseButton
        collapseSections={compareComponent.state.collapseSections}
        compareSetState={compareComponent.setState}
        sectionName='testSection'
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders the button collapsed if in collapsedSections', () => {
    compareComponent.state.collapseSections.testSection = true;
    const tree = shallow(
      <CollapseButton
        collapseSections={compareComponent.state.collapseSections}
        compareSetState={compareComponent.setState}
        sectionName='testSection'
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('collapses given section when clicked', () => {
    const tree = shallow(
      <CollapseButton
        collapseSections={compareComponent.state.collapseSections}
        compareSetState={compareComponent.setState}
        sectionName='testSection'
      />,
    );
    tree.find('WithStyles(Button)').simulate('click');
    expect(compareComponent.setState).toHaveBeenCalledWith({
      collapseSections: { testSection: true },
    });
  });

  it('expands given section when clicked if it is collapsed', () => {
    compareComponent.state.collapseSections.testSection = true;
    const tree = shallow(
      <CollapseButton
        collapseSections={compareComponent.state.collapseSections}
        compareSetState={compareComponent.setState}
        sectionName='testSection'
      />,
    );
    tree.find('WithStyles(Button)').simulate('click');
    expect(compareComponent.setState).toHaveBeenCalledWith({
      collapseSections: { testSection: false },
    });
  });
});
