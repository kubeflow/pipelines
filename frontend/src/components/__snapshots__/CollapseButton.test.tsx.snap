// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapseButton initial render 1`] = `
<div>
  <WithStyles(Button)
    className="collapseBtn"
    onClick={[Function]}
    title="Expand/Collapse this section"
  >
    <pure(ArrowDropUpIcon)
      className="icon"
    />
    testSection
  </WithStyles(Button)>
</div>
`;

exports[`CollapseButton renders the button collapsed if in collapsedSections 1`] = `
<div>
  <WithStyles(Button)
    className="collapseBtn"
    onClick={[Function]}
    title="Expand/Collapse this section"
  >
    <pure(ArrowDropUpIcon)
      className="icon collapsed"
    />
    testSection
  </WithStyles(Button)>
</div>
`;
